### PR TITLE
[Lim Fang Jie] iP

### PR DESCRIPTION
**Pull request by xzynos**

### CODE REVIEW

File: src/main/java/ConsoleInput.java [hunk 1]
```diff
@@ -0,0 +1,17 @@
+public class ConsoleInput {
```
Comment (okkhoy) [inline src/main/java/ConsoleInput.java]: Non trivial classes & methods need javadoc comments

File: src/main/java/ConsoleInterface.java [hunk 2]
```diff
@@ -0,0 +1,201 @@
+import java.util.Scanner;
+
+public class ConsoleInterface {
+    private static final String COMMAND_BYE = "bye";
+    private static final String COMMAND_LIST = "list";
+    private static final String COMMAND_MARK = "mark";
+    private static final String COMMAND_UNMARK = "unmark";
+    private static final String COMMAND_TODO = "todo";
+    private static final String COMMAND_DEADLINE = "deadline";
+    private static final String COMMAND_EVENT = "event";
```
Comment (okkhoy) [inline src/main/java/ConsoleInterface.java]: Great! You have followed the correct naming convention

File: src/main/java/ConsoleInterface.java [hunk 3]
```diff
@@ -0,0 +1,201 @@
...
+     * Reads user input from standard in.
+     *
+     * @return Command and arguments from the user.
+     */
+    public ConsoleInput getConsoleInput() {
+        String userInputStr = scanner.nextLine();
+        String[] userInputArr = userInputStr.split(" ", 2);
+        String command = userInputArr[0];
+        String argument = "";
+        if (userInputArr.length == 2) {
```
Comment (okkhoy) [inline src/main/java/ConsoleInterface.java]: Do I see a magic number here?

File: src/main/java/Deadline.java [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+        this.by = by;
+    }
+
+    public void print() {
+        String printString = "[D][";
+        if (super.isComplete) {
+            printString += "X";
+        } else {
+            printString += " ";
+        }
```
Comment (okkhoy) [inline src/main/java/Deadline.java]: This could be a part of  the base class

File: text-ui-test/input.txt [hunk 5]
```diff
@@ -0,0 +1,11 @@
...
+deadline return book /by Sunday
+event project meeting /at Mon 2-4pm
+list
+mark 1
+mark 2
+mark 3
+list
+deadline do homework /by no idea :-p
+list
+bye
```
Comment (okkhoy) [inline text-ui-test/input.txt]: Few more tests please!

**Review by okkhoy — COMMENTED** [review]
Pretty neat code. Good job!!

File: src/main/java/ConsoleInterface.java [hunk 6]
```diff
@@ -0,0 +1,201 @@
...
+                executeCommandEvent(input.getArguments());
+                break;
+            default:
+                break;
+            }
+
+            ConsoleInterface.printLineSeparator();
+        }
+    }
+}
```
Comment (jeyvia) [inline src/main/java/ConsoleInterface.java]: Very organised "main" java file to direct to other java files to task, and to print the respective commands! 😁

File: src/main/java/Deadline.java [hunk 7]
```diff
@@ -0,0 +1,19 @@
+public class Deadline extends Task {
+    protected String by;
+
+    public Deadline(String description, String by) {
+        super(description);
+        this.by = by;
+    }
+
+    public void print() {
```
Comment (jeyvia) [inline src/main/java/Deadline.java]: I think you can include this in ConsoleInterface.java, with the rest of the print functions

File: text-ui-test/EXPECTED.TXT [hunk 8]
```diff
@@ -5,3 +5,72 @@ Hello from
...
+Here are the tasks in your list:
+1.[T][X] borrow book
+2.[D][X] return book (by: Sunday)
+3.[E][X] project meeting (at: Mon 2-4pm)
+4.[D][ ] do homework (by: no idea :-p)
+____________________________________________________________
+
+____________________________________________________________
+Bye. Hope to see you again soon!
+____________________________________________________________
```
Comment (jeyvia) [inline text-ui-test/EXPECTED.TXT]: I think you can have more test cases to test your corner cases! 😎

File: src/main/java/Task.java [hunk 9]
```diff
@@ -0,0 +1,27 @@
+public abstract class Task {
+    protected String description;
+    protected boolean isComplete;
```
Comment (jeyvia) [inline src/main/java/Task.java]: Nice following naming conventions! 😊

**Review by jeyvia — COMMENTED** [review]
Overall, very good and neat code! Just a few minor changes :)

File: src/main/java/Duke.java [hunk 10]
```diff
@@ -1,10 +1,19 @@
...
+        ConsoleInterface.printLogo();
+        ConsoleInterface.printLineSeparator();
+        ConsoleInterface.printGreetingMessage();
+        ConsoleInterface.printLineSeparator();
+
+        consoleInterface.executeProgram();
+
+        ConsoleInterface.printLineSeparator();
+        ConsoleInterface.printGoodbyeMessage();
+        ConsoleInterface.printLineSeparator();
```
Comment (jingwei55) [inline src/main/java/Duke.java]: Methods are all using the correct camelCase, well done!

File: src/main/java/ConsoleInterface.java [hunk 11]
```diff
@@ -0,0 +1,201 @@
...
+    public static void printLogo() {
+        String logo = " ____        _        \n"
+                + "|  _ \\ _   _| | _____ \n"
+                + "| | | | | | | |/ / _ \\\n"
+                + "| |_| | |_| |   <  __/\n"
+                + "|____/ \\__,_|_|\\_\\___|\n";
+        System.out.println("Hello from\n" + logo);
+    }
+
+    /**
```
Comment (paullowse) [inline src/main/java/ConsoleInterface.java]: Do follow standard for javadoc comments

File: src/main/java/ConsoleInterface.java [hunk 12]
```diff
@@ -0,0 +1,201 @@
...
+     * Adds an event into task manager.
+     *
+     * @param arguments Raw arguments returned by the function {@link #getConsoleInput()}.
+     */
+    public void executeCommandEvent(String arguments) {
+        System.out.println("Got it. I've added this task:");
+
+        String[] argumentArray = arguments.split("/at");
+        String description = argumentArray[0].trim();
+        String at = argumentArray[1].trim();
```
Comment (jingwei55) [inline src/main/java/ConsoleInterface.java]: variable can be more specific

File: src/main/java/ConsoleInterface.java [hunk 13]
```diff
@@ -0,0 +1,201 @@
...
+
+            ConsoleInput input = getConsoleInput();
+
+            if (input.getCommand().equals(COMMAND_BYE)) {
+                break;
+            }
+
+            ConsoleInterface.printLineSeparator();
+
+            switch (input.getCommand()) {
```
Comment (paullowse) [inline src/main/java/ConsoleInterface.java]: Good use of switch case and statements

File: src/main/java/ConsoleInterface.java [hunk 14]
```diff
@@ -0,0 +1,201 @@
+import java.util.Scanner;
+
+public class ConsoleInterface {
```
Comment (paullowse) [inline src/main/java/ConsoleInterface.java]: Good use of naming conventions for static variables!

File: src/main/java/ConsoleInterface.java [hunk 15]
```diff
@@ -0,0 +1,201 @@
...
+    public static void printGreetingMessage() {
+        String greetingMessage = "Hello! I'm Duke\n"
+                + "What can I do for you?";
+        System.out.println(greetingMessage);
+    }
+
+    /**
+     * Prints goodbye message to standard out.
+     */
+    public static void printGoodbyeMessage() {
```
Comment (paullowse) [inline src/main/java/ConsoleInterface.java]: Good use of functions to facilitate the code

File: src/main/java/Duke.java [hunk 16]
```diff
@@ -1,10 +1,19 @@
...
-                + "| | | | | | | |/ / _ \\\n"
-                + "| |_| | |_| |   <  __/\n"
-                + "|____/ \\__,_|_|\\_\\___|\n";
-        System.out.println("Hello from\n" + logo);
+        ConsoleInterface consoleInterface = new ConsoleInterface();
+
+        ConsoleInterface.printLogo();
+        ConsoleInterface.printLineSeparator();
+        ConsoleInterface.printGreetingMessage();
+        ConsoleInterface.printLineSeparator();
```
Comment (jingwei55) [inline src/main/java/Duke.java]: Methods are clearly phrased and can be understood easily. Perhaps could combine all functions into a larger function to not show less important functions like printLineSeperator

File: src/main/java/ConsoleInterface.java [hunk 17]
```diff
@@ -0,0 +1,201 @@
...
+     * Marks a task as completed.
+     *
+     * @param taskNumberStr Raw arguments returned by the function {@link #getConsoleInput()}.
+     */
+    public void executeCommandMark(String taskNumberStr) {
+        int taskNumberInt = Integer.parseInt(taskNumberStr);
+        taskManager.markTaskAsCompleted(taskNumberInt);
+    }
+
+    /**
```
Comment (paullowse) [inline src/main/java/ConsoleInterface.java]: very neatly organized code!

**Review by paullowse — COMMENTED** [review]
Good job on the work done so far!

File: src/main/java/TaskManager.java [hunk 18]
```diff
@@ -0,0 +1,63 @@
...
+    /**
+     * Marks a task as completed.
+     *
+     * @param taskNumber Task number of task as shown by the function {@link #printTasks()}.
+     */
+    public void markTaskAsCompleted(int taskNumber) {
+        int index = taskNumber - 1;
+        tasks.get(index).setComplete(true);
+        String taskDescription = tasks.get(index).getDescription();
+        System.out.println("Nice! I've marked this task as done:");
```
Comment (jingwei55) [inline src/main/java/TaskManager.java]: Take note of indentations

**Review by jingwei55 — COMMENTED** [review]
Overall the coding structure is very good, just some minor adjustments need to be done. Methods and classes can all be easily understood. Well done!

File: src/main/java/TaskManager.java [hunk 19]
```diff
@@ -0,0 +1,63 @@
...
+
+    /**
+     * Marks a task as uncompleted.
+     *
+     * @param taskNumber Task number of task as shown by the function {@link #printTasks()}.
+     */
+    public void markTaskAsUncompleted(int taskNumber) {
+        int index = taskNumber - 1;
+        tasks.get(index).setComplete(false);
+        String taskDescription = tasks.get(index).getDescription();
```
Comment (jingwei55) [inline src/main/java/TaskManager.java]: Very clear function definitions! Code is really easy to read and understand, well done!

**Review by jingwei55 — COMMENTED** [review]
Add on to previous review

File: src/main/java/Todo.java [hunk 20]
```diff
@@ -0,0 +1,16 @@
+public class Todo extends Task {
+    public Todo(String description) {
+        super(description);
+    }
+
+    public void print() {
```
Comment (deveshl) [inline src/main/java/Todo.java]: The implementation of print() in Todo and Deadline is mostly similar, could possibly be moved to the base class

File: src/main/java/ConsoleInterface.java (hunk 2 above)
Comment (deveshl) [reply src/main/java/ConsoleInterface.java]: Using constants for commands is a good idea

File: src/main/java/Duke.java (hunk 16 above)
Comment (deveshl) [reply src/main/java/Duke.java]: As above - it may be worthwhile to have a function for printing text surrounded with line separators to decrease repetition

**Review by deveshl — COMMENTED** [review]
High quality code overall.

File: src/main/java/duke/Duke.java [hunk 21]
```diff
@@ -0,0 +1,27 @@
...
+
+import duke.userinterface.ConsoleInterface;
+
+/**
+ * Entrypoint of the Duke application.
+ */
+public class Duke {
+    /**
+     * Initializes application and starts interaction with user.
+     */
```
Comment (okkhoy) [inline src/main/java/duke/Duke.java:11]: 👍  for the header comments

File: src/main/java/duke/task/Task.java [hunk 22]
```diff
@@ -0,0 +1,259 @@
...
+        } else {
+            taskStr += "0";
+        }
+        taskStr += " | ";
+        taskStr += todo.description;
+
+        return taskStr;
+    }
+
+    private static String convertDeadlineToString(Deadline deadline) {
```
Comment (okkhoy) [inline src/main/java/duke/task/Task.java]: Why not use `toString` in the individual classes?

File: src/main/java/duke/task/Task.java [hunk 23]
```diff
@@ -0,0 +1,259 @@
...
+        } else if (type.equals("D")) {
+            String by = taskStrArr[3].trim();
+            return convertFromStringToDeadline(isCompleteStr, description, by);
+        } else if (type.equals("E")) {
+            String at = taskStrArr[3].trim();
+            return convertFromStringToEvent(isCompleteStr, description, at);
+        } else {
+            return convertFromStringToTask(isCompleteStr, description);
+        }
+    }
```
Comment (okkhoy) [inline src/main/java/duke/task/Task.java]: Good example of good SLAP 💯

File: src/main/java/duke/userinterface/ConsoleCommand.java [hunk 24]
```diff
@@ -0,0 +1,7 @@
+package duke.userinterface;
+
+/**
+ * Stores parsed arguments.
+ */
+public class ConsoleCommand {
```
Comment (okkhoy) [inline src/main/java/duke/userinterface/ConsoleCommand.java]: perhaps abstract class?

File: src/main/java/duke/userinterface/ConsoleInputParser.java [hunk 25]
```diff
@@ -0,0 +1,182 @@
...
+     * @return Parsed arguments for the corresponding commands.
+     * @throws ConsoleInputParserException.InvalidCommandMarkException     If the format of the mark command is not valid.
+     * @throws ConsoleInputParserException.InvalidCommandUnmarkException   If the format of the unmark command is not valid.
+     * @throws ConsoleInputParserException.InvalidCommandTodoException     If the format of the todo command is not valid.
+     * @throws ConsoleInputParserException.InvalidCommandDeadlineException If the format of the deadline command is not valid.
+     * @throws ConsoleInputParserException.InvalidCommandEventException    If the format of the event command is not valid.
+     * @throws ConsoleInputParserException.InvalidCommandDeleteException   If the format of the delete command is not valid.
+     * @throws ConsoleInputParserException.InvalidCommandFindException     If the format of the find command is not valid.
+     * @throws ConsoleInputParserException.CommandNotFoundException        If the command is not found.
+     */
```
Comment (okkhoy) [inline src/main/java/duke/userinterface/ConsoleInputParser.java]: 👍

File: src/main/java/duke/userinterface/ConsoleInterface.java [hunk 26]
```diff
@@ -0,0 +1,309 @@
+package duke.userinterface;
+
+import duke.task.*;
```
Comment (okkhoy) [inline src/main/java/duke/userinterface/ConsoleInterface.java]: please import the classes explicitly; don't use wildcard imports

File: src/main/java/duke/userinterface/ConsoleInterface.java [hunk 27]
```diff
@@ -0,0 +1,309 @@
...
+import java.time.LocalDateTime;
+import java.util.ArrayList;
+import java.util.Scanner;
+
+/**
+ * Provides functions to interface with user via standard input and standard output.
+ */
+public class ConsoleInterface {
+    public static final String TASKS_DIRECTORY_PATH = "./data/";
+    public static final String TASKS_FILENAME = "tasks.txt";
```
Comment (okkhoy) [inline src/main/java/duke/userinterface/ConsoleInterface.java]: A next step would be to use a config file to specify this rather than hardcoding the path and name.

**Review by okkhoy — COMMENTED** [review]
Excellent work! Please keep it up.
There are a few minor suggestions I have mentioned in the rest of the review, please take a look.
